0038-http-auth-profiles/http-auth-profiles.md

```diff
@@ -0,0 +1,191 @@
+
+This document defines common profiles that can be used to ensure implementations of Interledger software work securely in an interoperable fashion, while simultaneously allowing Interledger software operators to more clearly understand the trade-offs involved with various security choices.
+
+The profiles defined in this RFC should:
+
+* ...be as simple as possible while also being highly secure. 
+* ...clearly define any trade-offs involved in choosing a particular profile.
+* ...be easily implementable across programming languages in order to aid broad adoption.
+* ...utilise well-known and broadly deployed protocols in order to increase confidence in the profiles.
+* ...be suitable for high-performance use-cases, such as ILPv4 packet processing.
+* ...resist replays (emphasis: resist, not necessarily be replay-proof, in order to support high-performance use-cases).
+* ...not negate or prevent the ability of operators to layer-on more security "on top of" a profile defined in this RFC.
+
+## Specification
+
+### Overview
+This document defines three authentication profiles for securing HTTP network connections in Interledger.
+
+### Authentication
+When authenticating requests between Interledger nodes, it is important to choose an algorithm that maintains an appropriate balance between usability and security, while at the same time maintaining high-performance _and_ interoperability.
+
+In order to find this balance, this document defines three Authentication profiles, each with various trade-offs that should be considered before use:
+
+* `SIMPLE`: Allows two ILP nodes to utilize a previously agreed-upon shared-secret as a [Bearer token](https://tools.ietf.org/html/rfc6750) in all HTTP requests. Peers SHOULD consider this token to be opaque and SHOULD NOT derive any special meaning from the token. 
+
+* `JWT_HS_256`: Allows two ILP nodes to utilize a previously agreed-upon shared-secret in order to _derive_ a `Bearer token` that conforms to the JSON Web Token (JWT) specification as defined in [RFC-7519](https://www.rfc-editor.org/rfc/rfc7519.html) using the `HS_256` signing algorithm defined in section 3.2 of [RFC-7518](https://www.rfc-editor.org/rfc/rfc7518#section-3.2).
+
+* `JWT_RS_256`: Allows two ILP nodes to utilize public-key pairs to _derive_ a `Bearer token` that conforms to the JSON Web Token (JWT) specification as defined in [RFC-7519](https://www.rfc-editor.org/rfc/rfc7519.html) using the `RS_256` signing algorithm defined in section 3.3 of [RFC-7518](https://www.rfc-editor.org/rfc/rfc7518#section-3.3).
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests, but SHOULD support `JWT_HS_256` at a minimum. It is RECOMMENDED to use `JWT_RS_256` for production deployments.
```

Comment:
The advantages of JWT_RS_256 over JWT_HS_256 are kinda niche, and the performance hit is significant when you're potentially doing an RSA signature verification on every packet. Overall I think HS_256 is the better option. Also worth noting that JWT_HS_256/JWT_RS_256 is only more secure than SIMPLE when you're worried about your HTTPS connection being man-in-the-middled

Reply:
> the performance hit is significant when you're potentially doing an RSA signature verification on every packet

I think the idea is that you would combine a short token expiration (e.g., 5 mins) with a server-local cached authentication decision to improve performance.

Reply:
> Also worth noting that JWT_HS_256/JWT_RS_256 is only more secure than SIMPLE when you're worried about your HTTPS connection being man-in-the-middled

MITM is one worry, but there are plenty of other worries. E.g., logging of secrets, TLS termination (which is very common). Pretty much _any_ place that a token can be leaked expands the attack surface when tokens don't expire.

I'm uncomfortable with any auth system involving real money that doesn't expire it's auth tokens.

Reply:
To add to @sappenin's comment, another attack vector that can have its surface area reduced via expiry is replay attacks. I'm still not adept enough with ILP to eloquently articulate how that would happen, but it's a common concern with tokens that don't have expiration and/or aren't opaque.